msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/JsonHelper.java

```diff
@@ -123,90 +97,102 @@ static <T extends JsonSerializable<T>> T convertJsonStringToJsonSerializableObje
-     *        "values": [ clientCapabilities ]
-     *      }
-     *    }
-     *  }
-     */
     public static String formCapabilitiesJson(Set<String> clientCapabilities) {
-        if (clientCapabilities != null && !clientCapabilities.isEmpty()) {
-            ClaimsRequest cr = new ClaimsRequest();
-            RequestedClaimAdditionalInfo capabilitiesValues = new RequestedClaimAdditionalInfo(false, null, new ArrayList<>(clientCapabilities));
-            cr.requestClaimInAccessToken("xms_cc", capabilitiesValues);
-
-            return cr.formatAsJSONString();
-        } else {
+        if (clientCapabilities == null || clientCapabilities.isEmpty()) {
             return null;
         }
+
+        ClaimsRequest cr = new ClaimsRequest();
+        RequestedClaimAdditionalInfo capabilitiesValues = new RequestedClaimAdditionalInfo(
+                false, null, new ArrayList<>(clientCapabilities));
+        cr.requestClaimInAccessToken("xms_cc", capabilitiesValues);
+
+        return cr.formatAsJSONString();
     }
 
-    /**
-     * Merges given JSON strings into one Jackson JSONNode object, which is returned as a String
-     */
     static String mergeJSONString(String mainJsonString, String addJsonString) {
-        JsonNode mainJson;
-        JsonNode addJson;
-
         try {
-            mainJson = mapper.readTree(mainJsonString);
-            addJson = mapper.readTree(addJsonString);
+            Map<String, Object> mainMap = parseJsonToMap(mainJsonString);
+            Map<String, Object> addMap = parseJsonToMap(addJsonString);
+
+            mergeJsonMaps(mainMap, addMap);
+
+            return writeJsonMap(mainMap);
         } catch (IOException e) {
             throw new MsalClientException(e.getMessage(), AuthenticationErrorCode.INVALID_JSON);
         }
-
-        mergeJSONNode(mainJson, addJson);
-
-        return mainJson.toString();
     }
 
-    /**
-     * Merges given Jackson JsonNode object into another JsonNode
-     */
-    static void mergeJSONNode(JsonNode mainNode, JsonNode addNode) {
-        if (addNode == null) {
+    @SuppressWarnings("unchecked")
+    private static void mergeJsonMaps(Map<String, Object> mainMap, Map<String, Object> addMap) {
+        if (addMap == null) {
             return;
         }
 
-        Iterator<String> fieldNames = addNode.fieldNames();
-        while (fieldNames.hasNext()) {
+        for (Map.Entry<String, Object> entry : addMap.entrySet()) {
+            String key = entry.getKey();
+            Object value = entry.getValue();
 
-            String fieldName = fieldNames.next();
-            JsonNode jsonNode = mainNode.get(fieldName);
-
-            if (jsonNode != null && jsonNode.isObject()) {
-                mergeJSONNode(jsonNode, addNode.get(fieldName));
+            if (mainMap.containsKey(key) && mainMap.get(key) instanceof Map && value instanceof Map) {
+                mergeJsonMaps((Map<String, Object>) mainMap.get(key), (Map<String, Object>) value);
```

Comment:
Are all the inputs to this function controlled? Could malformed / malicious json be a problem here?

Reply:
Other than the recursive call, `mergeJsonMaps` is only used in the `mergeJSONString`, which converts the two json strings to a map using JsonReader so I don't think malformed json would be an issue.

Malicious json could be, as this is used for claims and client capabilities, which are ultimately strings that come from the customer. However, the values in those jsons aren't something we can control or really verify, so I don't think there's much more we can do beyond JsonReader's parsing.